src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.CompilationData.cs

```diff
@@ -70,27 +70,18 @@ public bool RemoveCachedSemanticModel(SyntaxTree tree)
                     return computeDeclarationAnalysisData();
                 }
 
-                DeclarationAnalysisData data;
                 lock (_declarationAnalysisDataMap)
                 {
-                    if (_declarationAnalysisDataMap.TryGetValue(declaration, out data))
+                    if (_declarationAnalysisDataMap.TryGetValue(declaration, out var cachedData))
                     {
-                        return data;
+                        return cachedData;
                     }
                 }
 
-                data = computeDeclarationAnalysisData();
-
+                DeclarationAnalysisData data = computeDeclarationAnalysisData();
                 lock (_declarationAnalysisDataMap)
                 {
-                    if (!_declarationAnalysisDataMap.TryGetValue(declaration, out DeclarationAnalysisData existingData))
-                    {
-                        _declarationAnalysisDataMap.Add(declaration, data);
-                    }
-                    else
-                    {
-                        data = existingData;
-                    }
+                    _declarationAnalysisDataMap[declaration] = data;
```

Comment:
❗️  The form from the previous commit easier to understand from a concurrency perspective. Use one of the following approaches instead:

1. The previous form (optionally create a `GetOrAdd` extension method if you need more clarity for the intended semantics)
2. Provide performance data indicating that the additional lookup is too costly to use, and to costly to otherwise fix
3. Provide an analysis that the `computeDeclarationAnalysisData` from end-to-end is deterministic